src/test/ui/consts/const_let_eq.rs

```diff
@@ -0,0 +1,470 @@
+    (arr.0)[0] &= 0b1011_1111_1111_1111;
+    d &= (arr.0)[0];
+    let mut arr: Bar<[u32; 1]> = Bar { x: [f] };
+    arr.x[0] &= 0b0111_1111_1111_1111;
+    d &= arr.x[0];
+
+    d
+}
+
+const fn bit_or_assign(W(a): W) -> u32 {
+    let f = 0b0000_0000_0000_0000;
+
+    // Mutables:
+    let mut d: u32 = a;
+    d |= 0b0000_0000_0000_0001;
+
+    // Array projection
+    let mut arr: [u32; 1] = [f];
+    arr[0] |= 0b0000_0000_0000_1001;
+    d |= arr[0];
+
+    // Field projection:
+    let mut foo: Foo<u32> = Foo(f);
+    let mut bar: Bar<u32> = Bar { x: f };
+    foo.0 |= 0b0000_0000_0001_0000;
+    bar.x |= 0b0000_0000_0100_0000;
+    d |= foo.0 | bar.x;
+
+    // Array + Field projection:
+    let mut arr: [Foo<u32>; 1] = [Foo(f)];
+    arr[0].0 |= 0b0000_0001_0000_0000;
+    d |= arr[0].0;
+    let mut arr: [Bar<u32>; 1] = [Bar { x: f }];
+    arr[0].x |= 0b0000_0010_0000_0000;
+    d |= arr[0].x;
+
+    // Field + Array projection:
+    let mut arr: Foo<[u32; 1]> = Foo([f]);
+    (arr.0)[0] |= 0b1000_0000_0000_0000;
+    d |= (arr.0)[0]; // /= 128
+    let mut arr: Bar<[u32; 1]> = Bar { x: [f] };
+    arr.x[0] |= 0b1100_0000_0000_0000;
+    d |= arr.x[0]; // /= 256
+
+    d
+}
+
+const fn bit_xor_assign(W(a): W) -> u32 {
+    let f = 0b0000_0000_0000_0000;
+
+    // Mutables:
+    let mut d: u32 = a;
+    d ^= 0b0000_0000_0000_0001;
+
+    // Array projection
+    let mut arr: [u32; 1] = [f];
+    arr[0] ^= 0b0000_0000_0000_0010;
+    d ^= arr[0];
+
+    // Field projection:
+    let mut foo: Foo<u32> = Foo(f);
+    let mut bar: Bar<u32> = Bar { x: f };
+    foo.0 ^= 0b0000_0000_0001_0000;
+    bar.x ^= 0b0000_0000_1000_0000;
+    d ^= foo.0 ^ bar.x;
+
+    // Array + Field projection:
+    let mut arr: [Foo<u32>; 1] = [Foo(f)];
+    arr[0].0 ^= 0b0000_0001_0000_0000;
+    d ^= arr[0].0;
+    let mut arr: [Bar<u32>; 1] = [Bar { x: f }];
+    arr[0].x ^= 0b0000_0010_0000_0000;
+    d ^= arr[0].x;
+
+    // Field + Array projection:
+    let mut arr: Foo<[u32; 1]> = Foo([f]);
+    (arr.0)[0] ^= 0b0100_0000_0000_0000;
+    d ^= (arr.0)[0];
+    let mut arr: Bar<[u32; 1]> = Bar { x: [f] };
+    arr.x[0] ^= 0b1000_0000_0000_0000;
+    d ^= arr.x[0];
+
+    d
+}
+
+macro_rules! test {
+    ($c:ident, $e:expr, $r:expr) => {
+        const $c: u32 = $e;
+        assert_eq!($c, $r);
+        assert_eq!($e, $r);
+    }
+}
+
+fn main() {
+    test!(BASICS, basics((2,)), 13);
+    test!(ADD, add_assign(W(1)), 10);
+    test!(MUL, mul_assign(A { a: 0 }), 256);
+    test!(DIV, div_assign([1]), 4);
+    test!(REM, rem_assign(W(5)), 5);
+    test!(SUB, sub_assign(W(8)), 0);
+    test!(SHL, shl_assign(W(1)), 0b1000_0000_0000_0000);
+    test!(SHR, shr_assign(W(256)), 1);
+    test!(AND, bit_and_assign(W(0b1011_1111_1111_1111_1111)), 0b0011_1100_1101_0100);
+    test!(OR, bit_or_assign(W(0b1011_0000_0000_0000)), 0b1111_0011_0101_1001);
+    test!(XOR, bit_xor_assign(W(0b0000_0000_0000_0000)), 0b1100_0011_1001_0011);
```

Comment:
Could you add a separate test file ensuring that these operators are *not* permitted for f32 & f64?

Reply:
There's no need to prevent operators for `f32` and `f64` since the type system forbids bit operations on floats. We have a lot of tests that check that already.

Reply:
AFAIK these tests reside in https://github.com/rust-lang/rust/blob/master/src/test/ui/consts/min_const_fn/min_const_fn.rs which demonstrate that operations are forbidden by using arithmetic operators; extending the tests to assignment operators is imo a good thing so that we can ensure that the assignment operators are also dealt with properly and to guard against future changes and bugs.

Reply:
Oh I thought you meant the bit operations specifically (`5f32 & 6f32` is not legal Rust anywhere).

Reply:
Haha no, though I see now how you might get that impression =P; I meant `+=`, `*=`, and so on.